pkg/agent/client/endpoint_resolver.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2024 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"reflect"
+	"sync/atomic"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/util/proxy"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+type Listener interface {
+	Enqueue()
+}
+
+type EndpointResolver struct {
+	name        string
+	namespace   string
+	serviceName string
+	servicePort int32
+	// serviceLister is used to retrieve the antrea Service when selecting one of the Service's Endpoints.
+	serviceLister       corev1listers.ServiceLister
+	serviceListerSynced cache.InformerSynced
+	// endpointLister is used to select and Endpoint for the antrea Service. This Endpoint will
+	// be used to access the Service without depending on the availability of a proxy (whether
+	// kube-proxy or AntreaProxy).
+	endpointsLister       corev1listers.EndpointsLister
+	endpointsListerSynced cache.InformerSynced
+	queue                 workqueue.RateLimitingInterface
+	listeners             []Listener
+	endpointURL           atomic.Pointer[url.URL]
+}
+
+func NewEndpointResolver(
+	namespace, serviceName string,
+	servicePort int32,
+	serviceInformer coreinformers.ServiceInformer,
+	endpointsInformer coreinformers.EndpointsInformer,
+) *EndpointResolver {
+	key := namespace + "/" + serviceName
+	controllerName := fmt.Sprintf("ServiceEndpointResolver-%s", key)
+	resolver := &EndpointResolver{
+		name:                  controllerName,
+		namespace:             namespace,
+		serviceName:           serviceName,
+		servicePort:           servicePort,
+		serviceLister:         serviceInformer.Lister(),
+		serviceListerSynced:   serviceInformer.Informer().HasSynced,
+		endpointsLister:       endpointsInformer.Lister(),
+		endpointsListerSynced: endpointsInformer.Informer().HasSynced,
```

Comment:
I'm a bit concerned it would start watching Endpoints API for all objects unconditionally, which was not the case before the patch. In the current main branch, only two components consume endpointsInformer: AntreaProxy wouldn't start it unless EndpointSlice API is not available; ServiceExternalIPController starts it but it's not enabled by default and perhaps we should make it favor EndpointSlice API in the future.

Maybe we could consider EndpointSlice API when it exists, or start a separate informer with a filter?